src/System.Private.CoreLib/shared/System/Range.cs

```diff
@@ -6,17 +6,54 @@
 
 namespace System
 {
+    /// <summary>Represent a range has start and end indexes.</summary>
+    /// <remarks>
+    /// Range is used by the C# compiler to support the range syntax.
+    /// <code>
+    /// int[] someArray = new int[5] { 1, 2, 3, 4, 5 };
+    /// int[] subArray1 = someArray[0..2]; // { 1, 2 }
+    /// int[] subArray2 = someArray[1..^0]; // { 2, 3, 4, 5 }
+    /// </code>
+    /// </remarks>
     public readonly struct Range : IEquatable<Range>
     {
+        /// <summary>Represent the inclusive start index of the Range.</summary>
         public Index Start { get; }
+
+        /// <summary>Represent the exclusive end index of the Range.</summary>
         public Index End { get; }
 
-        private Range(Index start, Index end)
+        /// <summary>Construct a Range object using the start and end indexes.</summary>
+        /// <param name="start">Represent the inclusive start index of the range.</param>
+        /// <param name="end">Represent the exclusive end index of the range.</param>
+        public Range(Index start, Index end)
         {
             Start = start;
             End = end;
         }
 
+        /// <summary>Destruct the range object according to a collection length and return the start offset from the beginning and the length of this range.</summary>
+        /// <param name="length">The length of the collection that the range will be used with. length has to be a positive value</param>
+        public (int, int) GetOffsetLength(int length)
```

Comment:
I thought we were avoiding using ValueTuple in public API.  We changed our minds?
cc: @terrajobst

Reply:
Yes this has been discussed and agreed on in the design meeting.

Reply:
I'm very surprised.  We went around and around and around on this, internally and publicly, in the discussion of Enumerable.Zip, but here it's no problem?

Reply:
I think we agreed on having the method, but I don't think we closed on whether or not to use tuples for this. We said that we should follow up on the tuple naming convention in a separate meeting. For this particular case, we concluded that we could achieve the same effect by using the `Deconstruct` method, such as:

```C#
public struct Range
{
    public OffsetAndLength GetOffsetAndLength()
    {
        return new OffsetAndLength(12, 3);
    }

    public readonly struct OffsetAndLength
    {
        public OffsetAndLength(int offset, int length)
        {
            Offset = offset;
            Length = length;
        }

        public int Offset { get; }
        public int Length { get; }

        public void Deconstruct(out int offset, out int length)
        {
            offset = Offset;
            length = Length;
        }
    }
}
```

The `Deconstruct` method allows this sort of deconstruction:

```C#
var range = ...
var (offset, length) = range.GetOffsetAndLength();
```

Reply:
Also can some summarize why return Tuple is a problem? I am not sure if the Destructor will cause any issue to have the JIT optimize it especially we are calling it in many places in our code now.

Reply:
> Also can someone summarize why return Tuple is a problem

In short, because the language team and us haven't agree on the naming convention. In general, tuples have limitations (like you can't increase the number of elements without breaking the ABI) but I don't think they apply here.